### ISSUE TITLE
Solo5 executables should not be executable by a host kernel

### ISSUE THREAD

**Issue by mato** [Solo5/solo5#442]
If a user attempts to run a Solo5 executable directly (regardless of target), the host kernel will in most cases attempt to load and run the executable.

This is undesirable for several reasons:

1. Bad UX for a "naive or beginning user": Running such an executable directly produces `Segmentation fault` on Linux and `Bus error (core dumped)` on FreeBSD.
2. Accidentally attempting to directly run a unikernel from an untrusted source is a potential security risk, since the host kernel's ELF loader runs in kernel space.

I considered several solutions:

1. Follow the ELF spec: set `e_ident[EI_OSABI]` to `ELFOSABI_STANDALONE`. Unfortunately the Linux (but not FreeBSD) ELF loader ignores these values entirely and will attempt to execute any ELF file with a matching processor architecture. Even if that were not the case, without providing our own Solo5-specific compiler toolchain we have no easy way to control these values in the resulting binary.

2. Provide a "stub": similarly to how PE executables have a DOS stub, that would be executed by the host kernel instead of the Solo5 start-up code. This is very hard, if at all possible, since tenders would have no way of distinguishing the "stub" `PT_LOAD` segment from the "real" `PT_LOAD` segments that doesn't involve inventing out-of-spec schemes to distinguish the two.

3. Provide a `PT_INTERP` pointing to a non-existent path: on Linux, this produces `No such file or directory` when attempting to directly run such a binary. On FreeBSD, this produces `ELF interpreter /nonexistent/solo5 not found, error 2`.

I'm going to go with option (3) as this is the least complex and best option overall. While the Linux response of `No such file or directory` is not ideal, it's still better than the current `Segmentation fault`.

**Comment by adamsteen** [comment]
@mato 
Have you just tried to run the .hvt or the .virtio?

I will test this on OpenBSD shortly!

**Comment by mato** [comment]
@adamsteen:

> Have you just tried to run the .hvt or the .virtio?

Yes, that's what I'm talking about... is the description not clear?

**TL;DR**: If you run a `.hvt` or `.virtio` or any other Solo5 target binary directly then in _most_[1] cases you will get `Segmentation fault` or similar from the host kernel. We don't want that.

I've known about this behaviour for ages, just have not had time to think about and try some solutions.

> I will test this on OpenBSD shortly!

Wait for a PR, coming up today, then test before/after.

[1] If your CPU architecture matches the binary then on Linux, always, on FreeBSD if the binary was also built on FreeBSD.

**Comment by cfcs** [comment]
I think it'd be good to fix, have two comments:
- With the current PR, if `/nonexistent/solo5` is a directory then this can have unintended side-effects. Can we consider making this empty or `/` instead? (At the very minimum we can perhaps have `/nonexistent/solo5/` (with a trailing slash to prevent actual loading), but this still means we will have unintuitive VFS operations.
- For at least the SPT target pointing to a system-wide installed SPT tender might down the line be nice UX, but that's probably a lot more work.

**Comment by mato** [comment]
@cfcs:

Thanks for the comments. I think you're overthinking it, but here goes...

> With the current PR, if /nonexistent/solo5 is a directory then this can have unintended side-effects.

Such as?

```
$ sudo mkdir -p /nonexistent/solo5
$ tests/test_hello/test_hello.hvt 
bash: tests/test_hello/test_hello.hvt: Permission denied
```

I don't see a problem. Having anything under `/nonexistent`, more so the directory `/nonexistent/solo5` seems about as likely as being hit by an asteroid in the next five minutes.

> Can we consider making this empty or / instead?

I considered making it empty, though that may be out-of-spec re the ELF specifications, but then decided on `/nonexistent/solo5` since that is "valid" and gives a more obvious clue when someone inspects the binary with `file` or `readelf -l`.

Another alternative would be `/dev/null`, which on Linux produces `Permission denied` regardless of whether or not the node at `/dev/null` is or is not executable. I've not tested that on the BSDs.

> For at least the SPT target pointing to a system-wide installed SPT tender might down the line be nice UX, but that's probably a lot more work.

_That_ is a completely different feature, and requires a ton more work which we're not ready for. Also, be aware that it'd result in the host kernel loading _both_ the interpreter and unikernel binary using its in-kernel ELF loader, which adds attack vectors we don't have at the moment...

**Comment by mato** [comment]
@cfcs:

Ah sorry, I misunderstood. You wrote:

> With the current PR, if /nonexistent/solo5 is a **directory** then this can have unintended side-effects.

... but meant to say "With the current PR, if /nonexistent/solo5 is a **file** ..."

In the _extremely_ unlikely case that `/nonexistent/solo5` is a file *and* a valid ELF executable, then yes, the kernel would attempt to load that as the interpreter and that is an unwanted side effect.

I'll add `/` as you suggested. For reference, here's what Linux and FreeBSD do in the various cases when using '/nonexistent/solo5/' for `PT_INTERP`:

```
# Linux
$ tests/test_hello/test_hello.hvt 
bash: tests/test_hello/test_hello.hvt: No such file or directory
$ sudo mkdir -p /nonexistent/solo5
$ tests/test_hello/test_hello.hvt 
bash: tests/test_hello/test_hello.hvt: Permission denied
$ sudo rmdir /nonexistent/solo5/; sudo cp ./hello-static /nonexistent/solo5
$ tests/test_hello/test_hello.hvt 
bash: tests/test_hello/test_hello.hvt: Not a directory
```

```
# FreeBSD 12
$ tests/test_hello/test_hello.hvt
ELF interpreter /nonexistent/solo5/ not found, error 2
Abort trap
$ sudo mkdir -p /nonexistent/solo5
$ tests/test_hello/test_hello.hvt
ELF interpreter /nonexistent/solo5/ not found, error 13
Abort trap
$ sudo rmdir /nonexistent/solo5; sudo cp ./hello-static /nonexistent/solo5
$ tests/test_hello/test_hello.hvt 
ELF interpreter /nonexistent/solo5/ not found, error 20
Abort trap
```

**Comment by adamsteen** [comment]
For OpenBSD Current

```
Before
$tests/test_hello/test_hello.hvt
tests/test_hello/test_hello.hvt[12]: Lp: not found
tests/test_hello/test_hello.hvt[12]: 44960p: not found
tests/test_hello/test_hello.hvt: syntax error: `(' unexpected
```
```
After
$tests/test_hello/test_hello.hvt
tests/test_hello/test_hello.hvt[1]: syntax error: `H$KF' unexpected
```

Not very clear on both accounts and couldn't tell you want any of those errors mean.

```
virtio
$tests/test_hello/test_hello.virtio
tests/test_hello/test_hello.virtio[34]: syntax error: `)' unexpected

muen
$tests/test_hello/test_hello.muen
tests/test_hello/test_hello.muen[1]: syntax error: `(' unexpected
```

**Comment by mato** [comment]
@adamsteen:

> Not very clear on both accounts and couldn't tell you want any of those errors mean.

I think I know what's happening there. Very old-school UNIX behaviour -- the kernel ELF loader is returning an error, but instead of failing then and there, the kernel then "helpfully" tries to pass the executable to `/bin/sh`. The errors you're seeing are coming from `/bin/sh` attempting to interpret the ELF as a shell script.

Nothing we can do about this OpenBSD behaviour, but it does show that (for whatever reason both with/without this change) OpenBSD does not execute the binary as an ELF executable.

Thanks for testing.

**Comment by adamsteen** [comment]
@mato, now that you mention it, I have seen and heard of this behaviour before!

**Comment by mato** [comment]
So, I spent several hours today trying to verify _what exactly_ is happening on OpenBSD. It turns out that my initial assessment of "It's a 1970s UNIX fallback" was correct, but my understanding of _why_ this was happening was not.

In the event that someone wants to attempt to get better behaviour on OpenBSD, here's what's going on at the moment (with and without #443):

1. User types `./test_hello.hvt` in a shell.
2. Shell does `execve("./test_hello.hvt", ...);`, the result of which is `ENOEXEC`. The reason for the `ENOEXEC` is _at least_ the absence of a `.note.openbsd.ident` in Solo5 binaries built on OpenBSD, though there are likely other dragons there.
3. Shell decides "Oh, it must be 1975 and #! has not been invented yet, let me help you!" and interprets `./test_hello.hvt` as a script, leading to nonsensical errors and miserable failure.

The reason #443 always works as intended on Linux and FreeBSD is that these systems perceive the binaries produced by #443 as "valid", get as far as attempting to load the ELF interpreter and thus fail out of `execve(...)` with an error **other than** `ENOEXEC`.

Implementing this behaviour on OpenBSD is left as an exercise to the reader. (Evil chuckle)

**Comment by mato** [comment]
#443 has been merged, but leaving this open in case someone wants to try their hand at a solution for OpenBSD.

**Comment by adamsteen** [comment]
@mato did a quick hack test by adding [os-note-elf.h](https://github.com/openbsd/src/blob/master/lib/csu/os-note-elf.h) to [bindings.h](https://github.com/Solo5/solo5/blob/master/bindings/bindings.h)

OpenBSD will produce
```$ ./test_hello.hvt
Abort trap
```

Not sure of an easy way of integrating that in tho, and is it worth it? or is there a better way?

**Comment by mato** [comment]
@adamsteen 

Thanks for the pointer. I figured that pretending to be an OpenBSD binary might squash the kernel fallback to /bin/sh. 

However:

#443 has two motivations -- one is better UX, the other is improving overall security posture by inducing the host kernel ELF loader to bail out _as early as possible_. It's unclear if the `Abort trap` you see is a result of the kernel bailing out early due to the non-existent `PT_INTERP` or if it's actually trying to _execute_ the code, i.e. successfully loading the binary, and then aborting on the code being "nonsense".

Can you verify if the behaviour you see with the OpenBSD ELF note added is the "early bailout" behaviour? Ping me if you need pointers on how to best do that.

Once we've verified that, then we can make a suitable change. I'll also need to add some more self-tests to cover this, as the ELF NOTE business is very subtle so need to be careful that adding the OpenBSD ELF note does not break any other host systems in unexpected ways.

**Comment by adamsteen** [comment]
I will see what i can do, this will probably a background task.

**Comment by mato** [comment]
@adamsteen 

> 
> I will see what i can do, this will probably a background task.

I have a work in progress in #452. Re-read OpenBSD [sys/kern/exec_elf.c](http://bxr.su/OpenBSD/sys/kern/exec_elf.c) _again_ over the weekend and it looks like this does indeed do what we want. The user-visible behaviour is `Abort trap` when attempting to exec a Solo5 binary, and I've verified with ktrace on 6.7 beta that the system appears to be trying to `NAMEI /nonexistent/solo5` so we're on the right code path.

**Comment by adamsteen** [comment]
@mato I tested #452 on my local machine and it works as you'd expect. Thank you
